# Курсовая работа № 2
## Аналитика. Начальный уровень

Курсовая работа состоит из двух частей – обязательной и дополнительной. **Для зачета необходимо выполнение только первой части.** Выполнение второй части может потребовать дополнительные знания Python.

- [Часть первая](#Часть-первая)
- [Часть вторая](#Часть-вторая)

## Часть первая


Перед вами стоит задача – подготовить аналитический отчет для HR-отдела. На основании проведенной аналитики предполагается составить рекомендации для отдела кадров по стратегии набора персонала, а также по взаимодействию с уже имеющимися сотрудниками.
<br><br> В базе данных лежит набор таблиц, которые содержат данные о сотрудниках вымышленной компании.
Сделайте обзор штата сотрудников компании. Составьте набор предметов исследования, а затем проверьте их на данных. Вся аналитика должна быть выполена с помощью SQL. Впоследствии данные можно визуализировать, однако финальные датафреймы для графиков также должны быть подготовлены с помощью SQL. <br><br>

Примеры гипотез:
1. Есть зависимость между `perfomance score` и тем, под чьим руководством работает сотрудник.
2. Есть зависимость между продолжительностью работы в компании и семейным положением сотрудника.
2. Есть зависимость между продолжительностью работы в компании и возрастом сотрудника.

<br><br>
Параметры для подключения следующие: хост – `dsstudents.skillbox.ru`, порт – `5432`, имя базы данных – `human_resources`, пользователь – `readonly`, пароль – `6hajV34RTQfmxhS`. Таблицы, доступные для анализа, – `hr_dataset`, `production_staff`, `recruiting_costs`, `salary_grid`.

In [1]:
import sqlalchemy
import psycopg2
import pandas as pd
import numpy as np

In [15]:
### YOUR CODE HERE ###
conn = 'postgresql+psycopg2://readonly:6hajV34RTQfmxhS@dsstudents.skillbox.ru:5432/human_resources'
engine = sqlalchemy.create_engine(conn)
conn = engine.connect()

Воспользуемся таблицами hr_dataset и production_staff. Выкачаем их полностью, чтобы ознакомиться с данными.

In [18]:
pd.read_sql("select * from hr_dataset", conn).head()

id         Employee Name  Employee Number  marriedid  maritalstatusid  \
0   1            Brown, Mia       1103024456          1                1   
1   2  LaRotonda, William         1106026572          0                2   
2   3      Steans, Tyrone         1302053333          0                0   
3   4       Howard, Estelle       1211050782          1                1   
4   5           Singh, Nan        1307059817          0                0   

   genderid  empstatus_id  deptid  perf_scoreid  age  ...  Date of Hire  \
0         0             1       1             3   30  ...    2008-10-27   
1         1             1       1             3   34  ...    2014-01-06   
2         1             1       1             3   31  ...    2014-09-29   
3         0             1       1             9   32  ...    2015-02-16   
4         0             1       1             9   30  ...    2015-05-01   

  Days Employed  Date of Termination       Reason For Term Employment Status  \
0          3317                 None  N/A - still employed            Active   
1          1420                 None  N/A - still employed            Active   
2          1154                 None  N/A - still employed            Active   
3            58           2015-04-15  N/A - still employed            Active   
4           940                 None  N/A - still employed            Active   

      department                  position        Manager Name  \
0  Admin Offices              Accountant I  Brandon R. LeBlanc   
1  Admin Offices              Accountant I  Brandon R. LeBlanc   
2  Admin Offices              Accountant I  Brandon R. LeBlanc   
3  Admin Offices  Administrative Assistant  Brandon R. LeBlanc   
4  Admin Offices  Administrative Assistant  Brandon R. LeBlanc   

          Employee Source         Performance Score  
0      Diversity Job Fair               Fully Meets  
1      Website Banner Ads               Fully Meets  
2         Internet Search               Fully Meets  
3  Pay Per Click - Google  N/A- too early to review  
4      Website Banner Ads  N/A- too early to review  

[5 rows x 29 columns]

In [20]:
pd.read_sql("select * from production_staff", conn).head()

id       Employee Name Race Desc Date of Hire    TermDate  \
0   1   Albert, Michael       White   2011-08-01        None   
1   2      Bozzi, Charles     Asian   2013-09-30  2014-08-07   
2   3  Butler, Webster  L     White   2016-01-28        None   
3   4         Dunn, Amy       White   2014-09-18        None   
4   5     Gray, Elijiah       White   2015-06-02        None   

        Reason for Term       Employment Status         Department  \
0  N/A - still employed                  Active  Production          
1              retiring  Voluntarily Terminated  Production          
2  N/A - still employed                  Active  Production          
3  N/A - still employed                  Active  Production          
4  N/A - still employed                  Active  Production          

             Position     Pay    Manager Name Performance Score  \
0  Production Manager  $54.50  Elisa Bramante       Fully Meets   
1  Production Manager  $50.50  Elisa Bramante       Fully Meets   
2  Production Manager  $55.00  Elisa Bramante           Exceeds   
3  Production Manager  $51.00  Elisa Bramante       Fully Meets   
4  Production Manager  $54.00  Elisa Bramante       Fully Meets   

   Abutments/Hour Wk 1  Abutments/Hour Wk 2  Daily Error Rate  \
0                  0.0                  0.0               0.0   
1                  0.0                  0.0               0.0   
2                  0.0                  0.0               0.0   
3                  0.0                  0.0               0.0   
4                  0.0                  0.0               0.0   

   90-day Complaints  
0                0.0  
1                0.0  
2                0.0  
3                0.0  
4                0.0

После попыток соединения по Employee Number - id, выходит, что таблицы связаны по Employee Name. Не самый лучший выбор, однако таблицы вообще составлены, мягко говоря, неконвенционально и не соответствуют похоже ни одной из нормальных форм:)

In [22]:
df_hr = pd.read_sql('select * from hr_dataset h inner join production_staff p on h."Employee Name" = p."Employee Name"', conn)

In [39]:
print(len(df_hr))
df_hr.head()

208


id       Employee Name  Employee Number  marriedid  maritalstatusid  \
0  11         King, Janet       1001495124          1                1   
1  54   Albert, Michael         1501072311          0                2   
2  55      Bozzi, Charles       1303054580          0                0   
3  56  Butler, Webster  L       1110029990          0                0   
4  57         Dunn, Amy         1409070147          0                0   

   genderid  empstatus_id  deptid  perf_scoreid  age  ...  \
0         0             1       2             3   63  ...   
1         1             1       5             3   49  ...   
2         1             5       5             3   48  ...   
3         1             1       5             4   34  ...   
4         0             1       5             3   44  ...   

        Employment Status         Department                 Position     Pay  \
0                  Active  Production         Production Technician I  $23.00   
1                  Active  Production              Production Manager  $54.50   
2  Voluntarily Terminated  Production              Production Manager  $50.50   
3                  Active  Production              Production Manager  $55.00   
4                  Active  Production              Production Manager  $51.00   

     Manager Name Performance Score Abutments/Hour Wk 1 Abutments/Hour Wk 2  \
0    Elijiah Gray       Fully Meets                 9.0                10.0   
1  Elisa Bramante       Fully Meets                 0.0                 0.0   
2  Elisa Bramante       Fully Meets                 0.0                 0.0   
3  Elisa Bramante           Exceeds                 0.0                 0.0   
4  Elisa Bramante       Fully Meets                 0.0                 0.0   

  Daily Error Rate 90-day Complaints  
0              1.0               1.0  
1              0.0               0.0  
2              0.0               0.0  
3              0.0               0.0  
4              0.0               0.0  

[5 rows x 45 columns]

Итак, у нас данные о 208 сотрудниках, из которых мы попробуем извлечь какие-то выводы.

### Гипотеза I: 
Женщины получают на одинаковой должности меньше, чем мужчины.

In [37]:
pd.read_sql('select h.genderid, h.position, avg(p."Pay"::numeric) as "avgpay", count(*) from hr_dataset h inner join production_staff p on h."Employee Name" = p."Employee Name" group by h.genderid, h.position order by position', conn)

genderid                      position     avgpay  count
0         0               President & CEO  23.000000      1
1         1            Production Manager  51.250000      8
2         0            Production Manager  47.416667      6
3         0       Production Technician I  19.131928     83
4         1       Production Technician I  19.061509     53
5         1      Production Technician II  25.381579     19
6         0      Production Technician II  25.378108     37
7         1  Software Engineering Manager  27.000000      1

Гипотеза не подтвердилась! Более того, на позиции Production Manager средняя почасовая оплата у женщин больше. Нельзя сказать, что такая статистика обусловлена выбросами, людей примерно поровну. Отмечаем, что президент компании получает на уровне техника.

### Гипотеза II: 
Женатые сотрудники в среднем старше, рассмотрим в разрезе полов

In [45]:
pd.read_sql('select h.genderid, h.marriedid, avg(h.age) as "avgage", count(*) from hr_dataset h inner join production_staff p on h."Employee Name" = p."Employee Name" group by h.genderid, h.marriedid order by h.genderid, h.marriedid', conn)

genderid  marriedid     avgage  count
0         0          0  40.039474     76
1         0          1  39.823529     51
2         1          0  38.634615     52
3         1          1  40.068966     29

In [47]:
#Средняя ошибка выборки для признакак
np.std(df_hr["age"])/np.sqrt(200)

0.6238057191617239